### PR TITLE
fix(rtmp): stop deleting a low frame rate input before its first media packet

### PR DESCRIPTION
**Pull request by dimiden**
## Summary

Gives an RTMP input a way to stay connected while it legitimately sends nothing between its `publish` command and its first media packet.
`PushProvider::ChannelTaskRunner()` treats silence of any length as a dead client, and before a stream publishes the packet silence timeout it judges by is whatever `PacketSilenceTimeoutMs` the operator set, and the channel creation default of 3000 ms when they set none, so a low frame rate encoder was deleted before its input stream was ever created.
`FirstMediaWaitTimeoutMs` sizes that window separately. It is off unless an operator sets it, so the default behaviour of that window does not change.
The rest of the change applies always: the connection is closed when such a channel is deleted, the time `PushStream::OnDataReceived()` spends is no longer charged as client silence, and `ChannelTaskRunner()` decides from one consistent read of the channel state.

## Motivation

`PushProvider::ChannelTaskRunner()` deletes a channel that has been silent for longer than its packet silence timeout.
Three things made that fire on healthy inputs.

1. Before a stream publishes, the packet silence timeout is the channel creation default of 3000 ms unless the operator set `PacketSilenceTimeoutMs`, which the RTMP providers already applied at that point. It measures silence of any RTMP byte, so in the window after `publish` it is really measuring control message silence. A captured OBS session at 1 fps sent nothing at all for 23.5 s after its `publish` command, and was deleted at 3 s with `Channel N is timed out, ~3000 ms elapsed since last received, deleting it`.
2. The timeout was also charged for time `PushStream::OnDataReceived()` spent inside OME. `AdmissionWebhooks` is a synchronous call from there and `<AdmissionWebhooks><Timeout>` has no upper bound, so a control server taking 6 s to answer got the caller's own channel deleted at 3 s.
3. A channel deleted before it joined an application was only erased from the channel map. Nothing closed its socket, so the encoder kept a connection it believed was still streaming and never retried.

## Changes

- Pre-publish window: `PushStream::ApplyConfiguredFirstMediaWaitTimeoutMs()` sizes the wait when the publish URL is validated, for both the legacy RTMP provider and E-RTMP. The option has no default. With it absent this falls through to `ApplyConfiguredPacketSilenceTimeoutMs()`, which is the call the RTMP providers made at that point before this change, and the wait never starts.
- End of the wait: with the option set, the first message carrying a coded frame hands the channel to an operator-set positive `PacketSilenceTimeoutMs`, and to the channel creation default otherwise. An explicit `0` disables the timeout only once the stream has actually published, through `PushApplication::JoinStream()`. The wait is a phase (`NotStarted`, `Waiting`, `Ended`), so a message that arrives before the wait was sized cannot consume it, and with the option absent there is no wait to consume at all.
- A codec description does not end the wait. An encoder can send the sequence header as soon as it has initialised: measured with ffmpeg at 1 fps, the AVC sequence header went out within 100 ms of `publish` and the first frame 65 s later, so ending the wait there left the option doing nothing. `pvd::rtmp::HasVideoFrame()` and `HasAudioFrame()` answer this, with one overload reading the legacy FLV tag header and one reading the values the enhanced parser produces. They name the layouts that hold a coded frame and answer no to everything else, a reserved value and a codec the path does not service included, because the two wrong answers do not cost the same: taking a description for a frame gets the channel deleted while the encoder is still filling its pipeline, while taking a frame for a description only keeps the operator's own wait running until publish. Both paths answer alike for legacy AVC and legacy AAC, which the unit tests pin; an enhanced tag header is a coded frame only on the E-RTMP path, because the legacy path cannot service those codecs at all.
- Transport teardown: `PushStream::CloseTransport()` closes the connection of a channel that is deleted before it joins an application, where `Application::DeleteStream()` and therefore `Stream::Stop()` never run.
- The time `PushStream::OnDataReceived()` spends is not client silence: the channel is marked while that call runs and `ChannelTaskRunner()` passes over it. `PushStream::BeginProcessingData()` and `EndProcessingData()` count rather than set a flag, because nothing in the base class guarantees that two calls for one channel never overlap.
- One consistent read: `PushStream::GetSilenceState()` reads everything `ChannelTaskRunner()` decides by in one pass, and `PushStream::TryBeginReaping()` commits the deletion with a single compare and swap over a packed word holding the `OnDataReceived()` count, a change counter and `REAPING_FLAG`. A caller that loses that swap writes nothing, so it cannot refuse a legitimate `OnDataReceived()` and drop its bytes.
- Admission redirect: the legacy provider stores the post-redirect application name when the publish URL is validated, which is where E-RTMP already stored it, so the first-media wait starts and ends against the same application, and the chunk parser's log prefix follows it from that point. `PacketSilenceTimeoutMs` already resolved against the redirected application on both paths and still does.
- Logging: `RtmpProvider::OnDisconnected()` no longer reports an error when OME closed the socket itself, which is the expected path after such a deletion.

## Configuration

`FirstMediaWaitTimeoutMs` is RTMP only, has no default, and must be positive. It is rejected at load time otherwise, including an empty or non numeric value, so a typo cannot silently remove the guard.
Leave it out and the pre-publish window behaves exactly as it did before this change.

```xml
<Applications>
  <Application>
    <Name>app</Name>
    <Providers>
      <RTMP>
        <!-- how long the input may stay silent between `publish` and its first media packet -->
        <!-- off unless set; positive only -->
        <FirstMediaWaitTimeoutMs>30000</FirstMediaWaitTimeoutMs>
        <!-- existing option: packet silence timeout, which also governs before the stream publishes -->
        <PacketSilenceTimeoutMs>5000</PacketSilenceTimeoutMs>
      </RTMP>
    </Providers>
  </Application>
</Applications>
```

The packet silence timeout an RTMP channel holds, by phase:

| phase                                                   | with `FirstMediaWaitTimeoutMs` set                                                     | with it absent                                                                         |
| ------------------------------------------------------- | -------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------- |
| connected, before `publish` is validated                | channel creation default, 3000 ms                                                      | same                                                                                   |
| `publish` validated, waiting for the first media packet | `FirstMediaWaitTimeoutMs`                                                              | operator-set positive `PacketSilenceTimeoutMs`, otherwise the channel creation default |
| first media packet received, not yet publishing         | operator-set positive `PacketSilenceTimeoutMs`, otherwise the channel creation default | unchanged from the row above                                                           |
| publishing                                              | the effective `PacketSilenceTimeoutMs`, `0` included, applied by `JoinStream()`        | same                                                                                   |

The middle two rows with the option absent are what the RTMP providers already did, so leaving the option out is not a behaviour change.

## Limitations

- A connection that sends no bytes at all is never deleted by `ChannelTaskRunner()`. `GetElapsedMsSinceLastReceived()` returns a negative value until the first accepted read, so the comparison never fires. This is unchanged from master and is left as a known issue for a separate task: starting the timeout at channel creation instead charges the socket worker's own blockage to whichever channel happens to be waiting behind it, which deletes healthy clients.
- While `PushStream::OnDataReceived()` is inside a channel, `ChannelTaskRunner()` passes over it with no upper bound. `<AdmissionWebhooks><Timeout>` maps to `SO_RCVTIMEO`, which applies per receive call while `HttpClient` reads in a loop, so a control server that trickles its response keeps that call inside the channel indefinitely. A call that never returns leaves its channel undeletable.
- The timeout measures silence of any RTMP byte, not of media specifically. A peer that keeps sending control messages OME ignores holds its connection open.
- Nothing widens the pre-publish window on an operator's behalf. They have to know either that `FirstMediaWaitTimeoutMs` exists, which is RTMP only, or that a positive `PacketSilenceTimeoutMs` already widens that window on every push provider.
- A track the RTMP path in use cannot service never ends the wait. On the legacy path that includes every enhanced tag header, so an input whose only track is, say, Opus audio holds the operator's wait until it publishes or disconnects rather than falling back to the shorter timeout.
- SRT, MPEG-TS and WebRTC gain no wait of their own, but the base-class changes reach them: `ChannelTaskRunner()` skips a channel while `OnDataReceived()` is inside it, deletion goes through `TryBeginReaping()`, and `CloseTransport()` runs for a channel deleted before it joined an application.

### Unit tests

Build with tests enabled, then:

- `ome_test_base --gtest_filter='PushStreamLifecycle.*'` (29 tests): the timeout at channel creation, the phase transitions of the first-media wait for an absent, zero and positive `PacketSilenceTimeoutMs`, a media message arriving before the wait was sized, `BeginProcessingData()` and `EndProcessingData()` across a normal, throwing and overlapping `OnDataReceived()`, the packed word invariant, exclusivity between `TryBeginReaping()` and `BeginProcessingData()`, a losing `TryBeginReaping()` writing nothing, and the transport teardown of a channel with and without an application.
- `ome_test_config --gtest_filter='ProviderTimeout.*'` (6 tests): the boundary values of both options and which of them the operator actually set.
- `ome_test_providers_rtmp --gtest_filter='RtmpMediaFrame.*'` (19 tests): which FLV tag layouts count as a coded frame, on both paths. Covers the AVC packet types including `EndOfSequence`, a video info command frame, an enhanced tag header on the legacy path, a video codec and a sound format the legacy path does not service, every enhanced audio and video packet type, and that both paths answer alike for legacy AVC and legacy AAC.

Full suite: 661 tests, 655 passing and 6 placeholder tests skipped.

### Manual A/B

The original report was reproduced by replaying a captured OBS session (1 fps, 23.5 s of silence after `publish`) byte for byte. Measured against that replay:

| build and configuration                          | result                                                                  |
| ------------------------------------------------ | ----------------------------------------------------------------------- |
| master                                           | deleted at 3 s, no input stream created                                 |
| this branch, option absent                       | deleted at 3 s, no input stream created (unchanged, as intended)        |
| this branch, `FirstMediaWaitTimeoutMs` 30000     | survives the 23.5 s, stream created                                     |
| this branch, `PacketSilenceTimeoutMs` 45000 only | survives, stream created (the pre-existing way of widening this window) |

The same replay was run against the E-RTMP path and against an admission webhook that redirects to another application.

The same delay is reproducible without a capture, because it is the encoder's frame pipeline: x264 holds `rc-lookahead + threads + bframes` frames before its first output, which is invisible at 60 fps and tens of seconds at 1 fps. Measured on this branch, the time from `publish` to the first frame follows that sum to within one frame (65.5 s for ffmpeg's own defaults, 24.5 s for the parameters below, 1.0 s with `threads=1`):

```bash
# about 24 s of encoder delay at 1 fps, which is what the captured OBS session showed
ffmpeg -re -f lavfi -i "testsrc=size=1280x720:rate=1" \
       -c:v libx264 -x264-params "threads=16:rc-lookahead=5:bframes=3" \
       -an -f flv rtmp://127.0.0.1:1935/app/t
```

With `FirstMediaWaitTimeoutMs` at 60000 the channel survives that; with the option absent it is deleted at 3 s. Note that ffmpeg sends the AVC sequence header immediately, so this input also covers the case where a description must not end the wait.

Enable trace logging for `PushProvider` so the timeout it judges by is visible:

```xml
<Tag name="PushProvider" level="trace" />
```

Publish a video only source, which defers publishing until enough messages have arrived, and freeze it once it is connected so the client goes silent without closing the connection:

```bash
ffmpeg -re -f lavfi -i "testsrc=size=320x240:rate=4" \
       -c:v libx264 -preset ultrafast -tune zerolatency -g 120 -an \
       -f flv rtmp://127.0.0.1:1935/app/vonly &
sleep 6
kill -STOP %1   # connected, timeout running, no more bytes
```

The log then shows the timeout the channel is judged by, and the elapsed value it is deleted at:

```
Checking channel 51, elapsed 9 ms, timeout 30000 ms      # waiting for the first media packet
Checking channel 51, elapsed 9 ms, timeout 3000 ms       # first media packet arrived
Channel 51 is timed out, 3102 ms elapsed since last received, deleting it
```

Measured with the source frozen after its first media packet:

| `FirstMediaWaitTimeoutMs` | `PacketSilenceTimeoutMs` | timeout transitions   | deleted at |
| ------------------------- | ------------------------ | --------------------- | ---------- |
| `30000`                   | unset                    | 3000 -> 30000 -> 3000 | 3102 ms    |
| `30000`                   | `0`                      | 3000 -> 30000 -> 3000 | 3066 ms    |
| `30000`                   | `2000`                   | 3000 -> 30000 -> 2000 | 2073 ms    |
| unset                     | unset                    | 3000 throughout       | 3079 ms    |

The last row is the option absent: no transition happens at all, which is the behaviour before this change.

For the webhook case, point `<AdmissionWebhooks><ControlServerUrl>` at a control server that answers `{"allowed": true}` after 6 s. On master the publishing channel is deleted at 3 s and no stream is created. On this branch the webhook completes and the stream is created, and a second client that connected before the worker blocked is served rather than deleted.

<details>
<summary>slow_webhook.py (control server used above)</summary>

```python
#!/usr/bin/env python3
"""Admission webhook control server that answers ALLOW after a delay."""
import json, sys, time
from http.server import BaseHTTPRequestHandler, HTTPServer

DELAY = float(sys.argv[1]) if len(sys.argv) > 1 else 6.0
PORT = int(sys.argv[2]) if len(sys.argv) > 2 else 3350


class H(BaseHTTPRequestHandler):
    def do_POST(self):
        self.rfile.read(int(self.headers.get("Content-Length", 0)))
        time.sleep(DELAY)
        body = json.dumps({"allowed": True}).encode()
        self.send_response(200)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    def log_message(self, *a):
        pass


HTTPServer(("127.0.0.1", PORT), H).serve_forever()
```

</details>

### CODE REVIEW

File: src/config/items/virtual_hosts/applications/providers/provider_timeout_test.cpp [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+//  Created by Hyunjun Jang
+//  Copyright (c) 2026 OvenMediaLabs. All rights reserved.
+//
+//==============================================================================
+#include <config/config.h>
+#include <gtest/gtest.h>
+
+#include <cstdio>
+#include <fstream>
+
```
[the remark this answers is not in the document]
Comment (dimiden) [reply src/config/items/virtual_hosts/applications/providers/provider_timeout_test.cpp:16]: Fixed.

File: src/base/provider/push_provider/stream_lifecycle_test.cpp [hunk 2]
```diff
@@ -0,0 +1,847 @@
...
+#include <gtest/gtest.h>
+
+#include <atomic>
+#include <chrono>
+#include <cstdio>
+#include <fstream>
+#include <functional>
+#include <future>
+#include <thread>
+
```
[the remark this answers is not in the document]
Comment (dimiden) [reply src/base/provider/push_provider/stream_lifecycle_test.cpp:22]: Fixed.

File: src/base/provider/push_provider/stream.cpp [hunk 3]
```diff
@@ -81,16 +89,90 @@ namespace pvd
...
+		{
+			if (_activity_state.compare_exchange_weak(state, state + 1))
+			{
+				return true;
+			}
+		}
+
+		// `TryBeginReaping()` has succeeded on this channel, so there is nothing left to hand data to.
+		return false;
+	}
```
[the remark this answers is not in the document]
Comment (dimiden) [reply src/base/provider/push_provider/stream.cpp:133]: Fixed.

File: src/base/provider/push_provider/stream.cpp [hunk 4]
```diff
@@ -81,16 +89,98 @@ namespace pvd
...
+		SilenceState state;
+
+		// `_activity_state` is read first, so a write to any field below shows up as a change.
+		state.activity		= _activity_state.load();
+		state.is_processing = (state.activity & HANDLER_COUNT_MASK) != 0;
+		state.timeout_ms	= GetPacketSilenceTimeoutMs();
+		state.elapsed_ms	= GetElapsedMsSinceLastReceived();
+
+		return state;
+	}
```
[the remark this answers is not in the document]
Comment (dimiden) [reply src/base/provider/push_provider/stream.cpp:172]: Fixed.

File: src/base/provider/push_provider/stream.cpp [hunk 5]
```diff
@@ -135,6 +228,85 @@ namespace pvd
...
+		const auto provider_type = provider->GetProviderType();
+
+		bool is_configured		 = false;
+		const auto configured_ms = application->GetConfiguredFirstMediaWaitTimeoutMs(provider_type, &is_configured);
+
+		// With the option unset no wait runs, and `PacketSilenceTimeoutMs` governs this window instead.
+		// A `0` cannot get here: the config layer rejects it, as it does an empty or non-numeric value.
+		if ((is_configured == false) || (configured_ms <= 0))
+		{
+			ApplyConfiguredPacketSilenceTimeoutMs(vhost_app_name);
```
[the remark this answers is not in the document]
Comment (dimiden) [reply src/base/provider/push_provider/stream.cpp:255]: Fixed.

**Review by Keukhan — COMMENTED** [review]
## 1. `logtt` → `logtd` in `ChannelTaskRunner()` looks unintentional

`src/base/provider/push_provider/provider.cpp`

The `Checking channel ...` line moved from trace to debug. That loop runs every 100 ms, once per channel, so at debug level it is 10 lines/s **per channel** — 1,000 lines/s at 100 channels, 5,000 at 500. `trace` is rarely on in production but `debug` frequently is, and this one line will bury everything else in it.

The PR description also tells operators to enable `<Tag name="PushProvider" level="trace" />` to see this exact log, so the code and the docs now disagree.

Suggest reverting it to `logtt`.

## 2. `CloseTransport()` is RTMP-only, so motivation #3 is still live for SRT

Motivation #3 — a channel deleted before it joins an application has nothing close its socket, so the encoder keeps a connection it believes is still streaming and never retries — is not RTMP-specific. It applies to every connection-oriented push provider. But the override only landed on the two RTMP streams:

| stream | `CloseTransport()` | on a pre-publish reap |
| --- | --- | --- |
| `RtmpStream` | overridden | socket closed |
| `RtmpStreamV2` | overridden | socket closed |
| `MpegTsStream` (used by **SRT** and MPEG-TS) | not overridden | base no-op, nothing happens |

SRT reaches that path easily: the channel is created at connect, `DoesBelongApplication()` stays false until the TS demuxer has resolved its tracks, and `SrtProvider::OnTimedOut()` is empty, so nothing else compensates. An SRT input reaped in that window leaves the encoder holding a live connection — exactly the symptom this PR fixes for RTMP.

The correct body already exists in `MpegTsStream::Stop()` (`mpegts_stream.cpp:108`):

```cpp
void MpegTsStream::CloseTransport()
{
    // SRT owns a connected socket. MPEG-TS over UDP shares a bound socket,
    // so the guard makes this a no-op there.
    if ((_remote != nullptr) && (_remote->GetState() == ov::SocketState::Connected))
    {
        _remote->Close();
    }
}
```

The `Connected` check is what makes it safe for both: the SRT socket is `Connected` and gets closed, the MPEG-TS UDP socket is `Bound` and is skipped — which matters, because that socket is shared across channels and must not be closed. Five lines covers both providers.

**Comment by dimiden** [comment]
@Keukhan 
1. I considered updating the PR description since changing `trace` to `debug` was originally intended to check if this issue persists in production. However, I went with your suggestion and changed it back to `trace`.
2. This PR is for RTMP. (Please check the PR title.)
3. This is a limitation of the current implementation. (Since addressing every edge case is impossible, it was discussed and agreed upon to wrap it up at the current implementation level.)

**Review by Keukhan — APPROVED** [review]
lgtm

**Comment by dimiden** [comment]
@naanlizard 
Could you check if the symptom you mentioned in [#2262](https://github.com/OvenMediaLabs/OvenMediaEngine/pull/2262#issuecomment-5089017874) is mitigated by the `FirstMediaWaitTimeoutMs` option?
If my assumption is correct, I believe this option should alleviate most of the cases you're encountering.

For your information, this is currently in an experimental phase to see if it can cover various streaming issues, including the one you reported and other internally logged issues, so I plan to keep it undocumented for the time being.

**Comment by naanlizard** [comment]
> @naanlizard Could you check if the symptom you mentioned in [#2262](https://github.com/OvenMediaLabs/OvenMediaEngine/pull/2262#issuecomment-5089017874) is mitigated by the `FirstMediaWaitTimeoutMs` option? If my assumption is correct, I believe this option should alleviate most of the cases you're encountering.
> 
> For your information, this is currently in an experimental phase to see if it can cover various streaming issues, including the one you reported and other internally logged issues, so I plan to keep it undocumented for the time being.


Tested on master (3f790904) with OBS in the configuration I reported ("Automatically record when streaming", Recording Quality "Same as stream"), with the RTMP handshake delayed so the streaming output attaches after the shared encoder's first IDR. 

No setting set, it at least tells OBS it failed to connect. Once I set `FirstMediaWaitTimeoutMs=30000`, the stream is able to successfully connect and play (though it takes a bit, 8s or so with default OBS settings). 

I'd suggest we make the default 30s for this new setting, or people upgrading from the previous release (which didn't have this issue) will run into suddenly failing streams and not know why

**Comment by dimiden** [comment]
@naanlizard
Thanks for testing it. The behaviour you saw is the fix for the case you described in #2262, where the client was never disconnected so OBS kept reporting everything as fine while the stream never played. It is working as intended.

The reason I want to keep this option opt-in is that turning it on by default changes behaviour that has been in place for a long time, and I cannot settle on the right value yet.

To go into more detail, the symptom you hit was not introduced by this change.
`ChannelTaskRunner()` and the 3000 ms default applied at channel creation came in with d28da8587, and have been part of releases since v0.18.3.0.
That code checked every channel without distinguishing whether it had published, so a channel that has not published yet has been subject to deletion since then.
Nothing in these changes shortened that window.
#2262 made an operator-set `PacketSilenceTimeoutMs` apply to the pre-publish window as well, and with nothing configured that window is still 3000 ms.
This PR does not make that deletion any earlier, and it actually skips a channel while `OnDataReceived()` is running.
What changed is that the socket is closed after the channel is deleted (previously it was left open, which is why the client still showed as connected).

So a default of 30 s, or any other value I might pick, only covers some of the cases.
(As measured in the PR description, at 1 fps the time from `publish` to the first frame is 65.5 s with FFmpeg's own defaults, so `30000` does not cover it either.)
And simply raising the value further has a side effect: a connection that sends `publish` and then no media at all stays around that much longer.
I would rather decide the default carefully, once I have seen more cases.

**Comment by naanlizard** [comment]
> The reason I want to keep this option opt-in is that turning it on by default changes behaviour that has been in place for a long time, and I cannot settle on the right value yet.
> 
> To go into more detail, the symptom you hit was not introduced by this change. `ChannelTaskRunner()` and the 3000 ms default applied at channel creation came in with [d28da85](https://github.com/OvenMediaLabs/OvenMediaEngine/commit/d28da8587bbf134f49f9e743ebdbfa23c8d1cc8c), and have been part of releases since v0.18.3.0. 

This is incorrect. We did not have the issue (as described by me previously, stream + recording using the same encoder leading to dropped streams) until upgrading to master FROM v0.20.5. Maybe that code existed, but it wasn't doing anything. 

I just verified this myself

**Comment by dimiden** [comment]
@naanlizard
You are right. I checked again, and this was not the original behaviour but a regression from #2068 (cf3df2e09).

Before that commit the silence timer was started only at publish, so a channel waiting to publish could never be deleted. #2068 made it start at the first received byte, as an unintended side effect of a data race fix.

So I will restore the previous behaviour rather than pick a default, and keep `FirstMediaWaitTimeoutMs` as an opt-in bound. Thanks for pushing back on it.
